File: ontrack-web-core/components/dashboards/widgets/project/LastActiveBranchesWidget.js
import {gql} from "graphql-request";
import {Space} from "antd";
import {useContext, useState} from "react";
import BranchBox from "@components/branches/BranchBox";
import SimpleWidget from "@components/dashboards/widgets/SimpleWidget";
import {DashboardContext} from "@components/dashboards/DashboardContext";
import LastActiveBranchesWidgetForm from "@components/dashboards/widgets/project/LastActiveBranchesWidgetForm";

export default function LastActiveBranchesWidget({count}) {

    const dashboard = useContext(DashboardContext)
    const [branches, setBranches] = useState([])

    // const check = checkContextIs('Last active branches', 'project')
    // if (check) return check

    return (
        <>
            <SimpleWidget
                title={`Last ${count} active branches`}
                query={
                    gql`
                        query LastActiveBranches(
                            $projectId: Int!,
                            $count: Int! = 10,
                        ) {
                            projects(id: $projectId) {
                                branches(count: $count, order: true) {
                                    id
                                    name
                                }
                            }
                        }
                    `
                }
                variables={{
                    projectId: dashboard.contextId,
                }}
                setData={data => {
                    setBranches(data.projects[0].branches)
                }}
                form={<LastActiveBranchesWidgetForm count={count}/>}
            >
                <Space direction="horizontal" size={16} wrap>
                    {branches.map(branch => <BranchBox key={branch.id} branch={branch}/>)}
                </Space>
            </SimpleWidget>
        </>
    )
}

File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/dashboards/widgets/project/LastActiveBranchesWidget.kt
package net.nemerosa.ontrack.model.dashboards.widgets.project

import net.nemerosa.ontrack.json.asJson
import net.nemerosa.ontrack.model.dashboards.WidgetInstance
import net.nemerosa.ontrack.model.dashboards.widgets.AbstractWidget
import net.nemerosa.ontrack.model.dashboards.widgets.WidgetConfig
import org.springframework.stereotype.Component

@Component
class LastActiveBranchesWidget : AbstractWidget<LastActiveBranchesWidget.LastActiveBranchesWidgetConfig>(
    key = "project/LastActiveBranches",
    name = "Last active branches"
) {
    data class LastActiveBranchesWidgetConfig(
        val count: Int = 10,
    ) : WidgetConfig

    fun toInstance(count: Int = 10) = WidgetInstance(
        "0",
        key,
        LastActiveBranchesWidgetConfig(count).asJson()
    )
}